orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java

```diff
@@ -0,0 +1,232 @@
+    webHookOutput.put("statusCode", e.getStatusCode());
+    webHookOutput.put("statusCodeValue", e.getStatusCode().value());
+    if (!StringUtils.isEmpty(e.getResponseBodyAsString())) {
+      webHookOutput.put("body", processResponseBodyAsJson(e.getResponseBodyAsString()));
+    }
+    TaskResult result = processReceivedFailureStatusCode(e.getStatusCode(), stageOutput);
+    log.warn(webHookOutput.get("error").toString(), e);
+    return result;
+  }
+
+  TaskResult processReceivedFailureStatusCode(HttpStatus status, Map<String, Object> stageOutput) {
+    String errorMessage;
+    ExecutionStatus executionStatus;
+    Map<String, Object> webHookOutput = (Map<String, Object>) stageOutput.get("webhook");
+    // Fail fast status check, retry status check or fail permanently
+    if (stageData.failFastStatusCodes != null
+        && stageData.failFastStatusCodes.contains(status.value())) {
+      errorMessage =
+          String.format(
+              "Received a status code %s configured to fail fast, terminating stage for pipeline %s to %s",
+              status.value(), executionId, stageData.url);
+      executionStatus = ExecutionStatus.TERMINAL;
+    } else if (status.is5xxServerError()
+        || IntStream.of(webhookProperties.getDefaultRetryStatusCodes())
+            .anyMatch(i -> i == status.value())) {
+      errorMessage =
+          String.format(
+              "Error submitting webhook for pipeline %s to %s with status code %s, will retry.",
+              executionId, stageData.url, status.value());
+      executionStatus = ExecutionStatus.RUNNING;
+    } else {
+      errorMessage =
+          String.format(
+              "Error submitting webhook for pipeline %s to %s with status code %s.",
+              executionId, stageData.url, status.value());
+      executionStatus = ExecutionStatus.TERMINAL;
+    }
+
+    webHookOutput.put("error", errorMessage);
+    return TaskResult.builder(executionStatus).context(stageOutput).build();
+  }
+
+  TaskResult processClientException(Exception e) {
+    String errorMessage;
+    ExecutionStatus executionStatus;
+    if (e instanceof UnknownHostException || e.getCause() instanceof UnknownHostException) {
+      errorMessage =
+          String.format(
+              "Name resolution failure in webhook for pipeline %s to %s, will retry.",
+              executionId, stageData.url);
+      executionStatus = ExecutionStatus.RUNNING;
+    } else if (stageData.method == HttpMethod.GET
+        && (e instanceof SocketTimeoutException
+            || e.getCause() instanceof SocketTimeoutException)) {
+      errorMessage =
+          String.format(
+              "Socket timeout in webhook on GET request for pipeline %s to %s, will retry.",
+              executionId, stageData.url);
+      executionStatus = ExecutionStatus.RUNNING;
+    } else {
+      errorMessage =
+          String.format(
+              "An exception occurred for pipeline %s to %s. %s",
```

Comment:
While we're in here...

```suggestion
              "An exception occurred for pipeline %s performing a request to %s. %s",
```